### ISSUE TITLE
Prevent segfault in Matlab when attempting to access bad index

### ISSUE THREAD

**Issue by tkuchida** [opensim-org/opensim-core#1250]
Because Matlab is 1-indexed and C++ is 0-indexed, Matlab users are likely to make off-by-one errors (OBOEs) when indexing into OpenSim Tables, etc. Currently, Matlab just kerplodes with an unhelpful segfault message when attempting to access a bad index, as reported by @jimmyDunne:

![jamesmatlabcrash](https://cloud.githubusercontent.com/assets/4203505/18487629/d41a55d2-79aa-11e6-91ca-d0d8801288d5.png)

@aymanhab suggests throwing/catching exceptions to prevent :boom:. See #1249.

**Comment by tkuchida** [comment]
Try replicating @jimmyDunne's example (the `for` loop in opensimVec3TableToStruct.m, line 12) in the API: is an index-out-of-bounds exception thrown and just not being caught correctly in Matlab? cc/ @aymanhab

**Comment by carmichaelong** [comment]
Likely that SimTK::VectorView has bound-checking and non-bound checking operators. Check if there's an Exception thrown (might be Release vs Debug too)

**Comment by chrisdembia** [comment]
> is an index-out-of-bounds exception thrown and just not being caught correctly in Matlab?

This sounds likely. In general, if OpenSim throws an exception but SWIG/Java doesn't know about it, MATLAB will crash.


> Likely that SimTK::VectorView has bound-checking and non-bound checking operators. Check if there's an Exception thrown (might be Release vs Debug too)

The non-checking method is usually `operator[]`, which probably isn't wrapped or available in the wrapping. We probably don't want to make it available in MATLAB either.

Might be best to tell SWIG/Java about the exception using `SWIG_DECLARE_EXCEPTION`, as is done [here](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Simulation/Model/Model.h#L206).

**Comment by aymanhab** [comment]
@jimmyDunne What's the minimal piece of code to test this. I can't create a new SimTK::RowVector (you get them from something else) and I'd like to write a minimal piece of code to get it (either build a table, add to it live, or use files if necessary), please advise @klshrinidhi or @jimmyDunne Thanks.

**Comment by klshrinidhi** [comment]
I think it is possible to create an object of `RowVector` since it is being wrapped [here](https://github.com/opensim-org/opensim-core/blob/master/Bindings/simbody.i#L86).

My guess is that this is a *segfault* due to no range checking in non-debug modes as code [here in simbody](https://github.com/simbody/simbody/blob/3dd798950b7cd8ba83e9338f0875c9647c8d09d4/SimTKcommon/include/SimTKcommon/internal/ExceptionMacros.h#L138-L150).

The solution may be to define the macro `SimTK_KEEP_RANGECHECK` so the check happens all the time. This will allow us to catch it in upper layers.

**Comment by aymanhab** [comment]
RowVectorView has no constructor (even though it's wrapped) and neither does RowVectorViewVec3 that's why I wanted to know how to get one.  If these are not the correct classes please let me know. Indeed I suspect it's the no-range-check indexing but I'm trying to have an actual piece of code to add as a test for future reference. If we map the operators to methods (as we likely have to in SWIG interface) then we can map to the range-check operator if there is one.

**Comment by jimmyDunne** [comment]
I am trying to build a TimesSeriesTableVec3 from scratch. I assume you first have an empty `TimeSeriesTableVec3()` , fill a row object, then append the row to the TimeSeriesTable?

I can make a `RowVector()` , but then it is unclear how to fill this with Vec3's. I don't have access to `RowVectorViewVec3`, as Ayman also noted. 

How do you build a Table (or a Vec3 table) from scratch? 

```
new_table = TimeSeriesTableVec3()

rowvector = RowVector()

%% Add some numbers to the rowvector

new_table.appendRow(rowvector)

```

**Comment by jimmyDunne** [comment]
As far as I can tell, there is no way in Matlab to edit the values of a RowVector().

**Comment by aymanhab** [comment]
An example in C++ to begin with would be a good start @klshrinidhi I'm sure you have something in the repo but I don't know where to look. Thanks.

**Comment by chrisdembia** [comment]
https://github.com/opensim-org/opensim-core/tree/master/OpenSim/Examples/DataTable

**Comment by aymanhab** [comment]
Thanks @chrisdembia I created a branch to add the test that we can all refer to https://github.com/opensim-org/opensim-core/tree/test_tables_java

Haven't tested indexing out of bounds yet. Wanted to see this pass first.

**Comment by jimmyDunne** [comment]
This is how it looks in Matlab

```

import org.opensim.modeling.*

new_table = TimeSeriesTable()

table = DataTable

colnames = StdVectorString;

colnames.isEmpty

colnames.add('0')
colnames.add('1')
colnames.add('2')
colnames.add('3')
colnames.add('4')

colnames.isEmpty
colnames.size

table.setColumnLabels(colnames);


row0 = RowVector(5, 0);
row1 = RowVector(5, double(0.25) );
row2 = RowVector(5, double(0.50) );
row3 = RowVector(5, double(0.75) );
row4 = RowVector(5, double(1) );


table.appendRow(0, row0);
table.appendRow(0, row1);
table.appendRow(0, row2);
table.appendRow(0, row3);
table.appendRow(0, row4);

% Retrieve a column by its label.
col3 = table.getDependentColumn('3');

col3.getElt(0,0)
col3.getElt(1,0)
col3.getElt(2,0)
col3.getElt(3,0)
col3.getElt(4,0)


col2 = table.getDependentColumn('2');

col2.getElt(0,0)
col2.getElt(1,0)
col2.getElt(2,0)
col2.getElt(3,0)
col2.getElt(4,0)
``

**Comment by jimmyDunne** [comment]
Sorry `row1 = RowVector(5, 0.25 );` does work. I added the double for debugging but didn't get rid of it. 

The above code can only be used once. If I try to do a second pass, Matlab crashes.

**Comment by jimmyDunne** [comment]
example 2

```

import org.opensim.modeling.*

new_table = TimeSeriesTable()

table = DataTable

colnames = StdVectorString;

colnames.isEmpty

colnames.add('0')
colnames.add('1')
colnames.add('2')
colnames.add('3')
colnames.add('4')

colnames.isEmpty
colnames.size

table.setColumnLabels(colnames);


row0 = RowVector(5, 0);
row1 = RowVector(5, 3.24 );
row2 = RowVector(5, 4.6 );
row3 = RowVector(5, 9.2 );
row4 = RowVector(5, 1.112 );


table.appendRow(0, row0);
table.appendRow(0.25, row1);
table.appendRow(0.5, row2);
table.appendRow(0.75, row3);
table.appendRow(1, row4);

% Following construction of TimeSeriesTable succeeds because the DataTable
% has its independent column strictly increasing.
timeseriestable = TimeSeriesTable(table);

% Editing the DataTable to not have strictly increasing independent column
% will fail the construction of TimeSeriesTable.

table.appendRow(0.9,row3)

try
   timeseriestable = TimeSeriesTable(table);
catch exception
   'Times series table exception: Time has to be increasing'
end

% Edit the entry in the independent column to make the column strictly
% increasing.
table.setIndependentValueAtIndex(5, 1.25);

try
   timeseriestable = TimeSeriesTable(table);
catch exception
   'Times series table exception: Time has to be increasing'
end
```

**Comment by klshrinidhi** [comment]
I have a snippets to reproduce the problem for [C++](https://github.com/opensim-org/opensim-core/blob/issue_1250_rowvectorview_crash/OpenSim/Common/Test/testDataTable.cpp#L185-L209) and [Python](https://github.com/opensim-org/opensim-core/blob/issue_1250_rowvectorview_crash/Bindings/Python/tests/test_DataTable.py#L71-L80).

In C++, I did not a *segfault* on my machine. It just printed garbage as shown in the snippet.
In Python, there was an exception:
```
 Traceback (most recent call last):
   File "<path-to>/opensim-core/Bindings/Python/tests/test_DataTable.py", line 81, in test_DataTable
     print table.getRowAtIndex(0)[4]
   File "opensim/simbody.py", line 5661, in __getitem__
     return _simbody.RowVectorBaseDouble___getitem__(self, i)
RuntimeError: std::exception in 'double SimTK::RowVectorBase< double >::__getitem__(size_t)': 
```
The exception in Python is because `RowVectorBase` is [extended in simbody.i](https://github.com/opensim-org/opensim-core/blob/issue_1250_rowvectorview_crash/Bindings/simbody.i#L36-L50) to do so.

There may have to be a similar extension for Java. Java's native indexing is also 0 based.

**Comment by chrisdembia** [comment]
@klshrinidhi in Python, was the message "Index not in range" printed? See [here](https://github.com/opensim-org/opensim-core/blob/issue_1250_rowvectorview_crash/Bindings/Python/swig/python_common.i#L42):

```
%exception {
    try {
        $action
    } catch (const std::exception& e) {
        std::string str("std::exception in '$fulldecl': ");
        std::string what(e.what());
        SWIG_exception_fail(SWIG_RuntimeError, (str + what).c_str());
    }
}
```

The exception message should have been printed.

**Comment by klshrinidhi** [comment]
The CI builds -- [Travis](https://travis-ci.org/opensim-org/opensim-core/builds/161259752) & [Appveyor](https://ci.appveyor.com/project/opensim-org/opensim-core/build/4690) -- show the resulting exception. Take a look at those.

**Comment by aymanhab** [comment]
Nope @chrisdembia only the name of the method is shown. I vaguely remember exception messages getting lost during stack unwinding following a throw but we need to debug to see why. We also need to make sure this reproduces and can be caught in Java as well.

**Comment by aymanhab** [comment]
Test now crashes as expected 
https://travis-ci.org/opensim-org/opensim-core/builds/161450500
Build log shows:

terminate called after throwing an instance of 'SimTK::Exception::Base'
  what():  

The branch is called test_tables_java

**Comment by klshrinidhi** [comment]
Okay. I will take a look and get back.

**Comment by klshrinidhi** [comment]
Addressed by #1288. It prevents seg-fault by throwing exception.
